### PR TITLE
feat(compaction): add vnode watermark picker

### LINKED ISSUE #17157: Reclaim space more aggresively for table with vnode table watermark specificed (table with range delete)
After switching to vnode table watermark (#13148) for range delete, we have ensured that range deletion is done in a monotonical way. This means that the per vnode deletion watermark is either monotonically increasing or decreasing. With this property, we have simplified the range delete implementation in many places of the storage code paths but compaction strategy is still agnostic to the deletion watermark. 

The current status:
- Compaction strategy generates compaction task without considering the per vnode table watermark. Only LSM balancedness and point deletion tombstone ratio are considered.
- Compactor will take per vnode table watermark into consideration during the execution of compaction task. That means if a task includes SST falling below the vnode deletion watermark, the corresponding data will be reclaim after the compaction is done.

Potential issues with the current status:
- There is no bound on when the data deleted by range delete can be reclaim. It is possible that the space for such tables keep increasing even when the range deletion happens more frequently then writes.
- For append-only state table with range deletion (e.g. log store state table), trivial move compaction tasks can be the majority and almost all space cannot be reclaim in time in such cases.

Ideas:
- Given the per vnode table watermark is monotonic, we can have a simple picker to pick SSTs with key range fully covered by the watermark for space reclaim. These are trivial space reclaim tasks with no actual compaction execution involved.
- For simplicity, this picker can only take bottommost level into consideration.
- To extend from above idea, we can generalize the picker to pick SSTs with `>X%` key range covered by the watermark for space reclaim. This can generate actual compaction tasks to execute.

### PR DESCRIPTION
**Pull request by zwang28**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

close #17157

- Add a new picker that picks trivial reclaim compaction task. For simplicity, this picker only takes bottommost level into consideration.
- Extract `SkipWatermarkState::from_safe_epoch_watermarks` and `HummockVersion::safe_epoch_table_watermarks` into pure functions for easier reuse.
- Add a `CompactionSelectorContext` to convey parameters to compaction selectors.

## Checklist

- [ ] I have written necessary rustdoc comments
- [ ] I have added necessary unit tests and integration tests
- [ ] I have added test labels as necessary. See [details](https://github.com/risingwavelabs/risingwave/blob/main/docs/developer-guide.md#ci-labels-guide).
- [ ] I have added fuzzing tests or opened an issue to track them. (Optional, recommended for new SQL features #7934).
- [ ] My PR contains breaking changes. (If it deprecates some features, please create a tracking issue to remove them in the future).
- [ ] All checks passed in `./risedev check` (or alias, `./risedev c`)
- [ ] My PR changes performance-critical code. (Please run macro/micro-benchmarks and show the results.)

- [ ] My PR contains critical fixes that are necessary to be merged into the latest release. (Please check out the [details](https://github.com/risingwavelabs/risingwave/blob/main/CONTRIBUTING.md))

## Documentation

- [ ] My PR needs documentation updates. (Please use the **Release note** section below to summarize the impact on users)

## Release note

If this PR includes changes that directly affect users or other significant modifications relevant to the community, kindly draft a release note to provide a concise summary of these changes. Please prioritize highlighting the impact these changes will have on users.

</details>

### CODE REVIEW

File: src/storage/hummock_sdk/src/version.rs [hunk 1]
```diff
@@ -39,7 +39,7 @@ use crate::{CompactionGroupId, HummockSstableObjectId, HummockVersionId, FIRST_V
 
 #[derive(Debug, Clone, PartialEq)]
 pub struct HummockVersionStateTableInfo {
-    state_table_info: HashMap<TableId, PbStateTableInfo>,
+    pub state_table_info: HashMap<TableId, PbStateTableInfo>,
```
Comment (wenym1) [inline src/storage/hummock_sdk/src/version.rs:42]: It's intended to make it not `pub`. we can use `.info()` to access it.

**Comment by zwang28** [comment]
One issue is compaction tasks of dynamic type has the highest priority, potentially starving space reclaim tasks.

**Comment by hzxa21** [comment]
> One issue is compaction tasks of dynamic type has the highest priority, potentially starving space reclaim tasks.

The assumptions are:
1. dynamic task should be done ASAP because it will cause write stall
2. dynamic task will also reclaim tombstones/space if it can

2 is not 100% valid because dynamic task will not take reclaim releated statistics like tombstone ratio into consideration. Also, for trivial reclaim task, which uses very few resource, it can be done relatively fast and won't block dynamic tasks.

However, this is a separate issue pre-exists prior to this PR so we don't need to block this PR because of that.

**Review by hzxa21 — APPROVED** [review]
LGTM.

Do we have a integration test (like the one in compaction test pipeline maybe?) to verify the effectiveness of the new picker?

**Comment by zwang28** [comment]
> Do we have a integration test (like the one in compaction test pipeline maybe?) to verify the effectiveness of the new picker?

I'm still considering how to add the corresponding tests.

**Pull request merged**